web_monitoring/internetarchive.py

```diff
@@ -87,149 +92,187 @@ def cdx_hash(content):
-                timestamp=data.timestamp, url=data.url)
-        )
-        count += 1
-        yield data
-
-    return count
-
-
-def list_versions(url, *, from_date=None, to_date=None, skip_repeats=True):
+    session : :class:`requests.Session`, optional
     """
-    Search archive.org for captures of a URL (optionally, within a time span).
-
-    This function provides a convenient, use-case-specific interface to
-    archive.org's CDX API. For a more direct, low-level API, use
-    :func:`search_cdx`.
-
-    Note that even URLs without wildcards may return results with multiple
-    URLs. Search results are matched by url_key, which is a SURT-formatted,
-    canonicalized URL:
-
-      * Does not differentiate between HTTP and HTTPS
-      * Is not case-sensitive
-      * Treats `www.` and `www*.` subdomains the same as no subdomain at all
-
-    Parameters
-    ----------
-    url : string
-        The URL to list versions for. Can contain wildcards.
-    from_date : datetime, optional
-        Get versions captured after this date.
-    to_date : datetime, optional
-        Get versions captured before this date.
-    skip_repeats : boolean, optional
-        Don’t include consecutive captures of the same content (default: True).
-
-    Raises
-    ------
-    UnexpectedResponseFormat
-        If the CDX response was not parseable.
-    ValueError
-        If there were no versions of the given URL.
-
-    Examples
-    --------
-    Grab the datetime and URL of the version nasa.gov snapshot.
-    >>> versions = list_versions('nasa.gov')
-    >>> version = next(versions)
-    >>> version.date
-    datetime.datetime(1996, 12, 31, 23, 58, 47)
-    >>> version.raw_url
-    "http://web.archive.org/web/19961231235847id\_/http://www.nasa.gov:80/"
-
-    Loop through all the snapshots.
-    >>> for version in list_versions('nasa.gov'):
-    ...     # do something
-    """
-    params = {'url': url, 'collapse': 'digest'}
-    if from_date:
-        params['from'] = from_date.strftime(URL_DATE_FORMAT)
-    if to_date:
-        params['to'] = to_date.strftime(URL_DATE_FORMAT)
-
-    last_hashes = {}
-    for version in search_cdx(params):
-        # TODO: may want to follow redirects and resolve them in the future
-        if not skip_repeats or last_hashes.get(version.url) != version.digest:
-            last_hashes[version.url] = version.digest
-            # TODO: yield the whole version
-            yield version
-
-    if not last_hashes:
-        raise ValueError("Internet archive does not have archived "
-                         "versions of {}".format(url))
+    def __init__(self, session=None):
+        self.session = session or requests.Session()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def close(self):
+        "Close the client's session."
+        self.session.close()
+
+    def search(self, params):
```

Comment:
Now that we have our headers around what the important parameters are, I'd like to explode this function signature to expose them:

```py
def search(resolve_revisits, show_resume_key, **kwargs):
    ...
```

The `**kwargs` provides an escape hatch to pass through any params that we miss or may be added later before our library is updated. My goal here is to make it much more obvious to the caller what the allowed and important parameters are.

Reply:
Should we just add all the known options as explicit keywords, then?

Reply:
Yes, that's my preference. I even have [a rant about why](https://nsls-ii.github.io/scientific-python-cookiecutter/guiding-design-principles.html#complexity-is-always-conserved).